rclcpp/include/rclcpp/mapped_ring_buffer.hpp

```diff
@@ -0,0 +1,205 @@
+   *
+   * /param key the key associated with the stored value
+   * /param value if the key is found, the value is stored in this parameter
+   */
+  void
+  get_copy_at_key(uint64_t key, std::unique_ptr<T> & value)
+  {
+    auto it = get_iterator_of_key(key);
+    value = nullptr;
+    if (it != elements_.end() && it->in_use) {
+      value = std::unique_ptr<T>(new T(*it->value));
+    }
+  }
+
+  /// Return ownership of the value stored in the ring buffer, leaving a copy.
+  /* The key is matched if an element in the ring bufer has a matching key.
+   * This method will allocate in order to store a copy.
+   *
+   * The key is not guaranteed to be unique, see the class docs for more.
+   *
+   * The ownership of the currently stored object is returned, but a copy is
+   * made and stored in its place.
+   * This means that multiple calls to this function for a particular element
+   * will result in returning the copied and stored object not the original.
+   * This also means that later calls to pop_at_key will not return the
+   * originally stored object, since it was returned by the first call to this
+   * method.
+   *
+   * /param key the key associated with the stored value
+   * /param value if the key is found, the value is stored in this parameter
+   */
+  void
+  get_ownership_at_key(uint64_t key, std::unique_ptr<T> & value)
+  {
+    auto it = get_iterator_of_key(key);
+    value = nullptr;
+    if (it != elements_.end() && it->in_use) {
+      // Make a copy.
+      auto copy = std::unique_ptr<T>(new T(*it->value));
+      // Return the original.
+      value.swap(it->value);
+      // Store the copy.
+      it->value.swap(copy);
+    }
+  }
+
+  /// Return ownership of the value stored in the ring buffer at the given key.
+  /* The key is matched if an element in the ring buffer has a matching key.
+   *
+   * The key is not guaranteed to be unique, see the class docs for more.
+   *
+   * /param key the key associated with the stored value
+   * /param value if the key is found, the value is stored in this parameter
+   */
+  void
+  pop_at_key(uint64_t key, std::unique_ptr<T> & value)
+  {
+    auto it = get_iterator_of_key(key);
+    value = nullptr;
+    if (it != elements_.end() && it->in_use) {
+      value.swap(it->value);
+      it->in_use = false;
+    }
+  }
+
+  /// Insert a key-value pair, displacing an existing pair if necessary.
+  /* The key's uniqueness is not checked on insertion.
+   * It is up to the user to ensure the key is unique.
+   * This method should not allocate memory.
+   *
+   * After insertion, if a pair was replaced, then value will contain ownership
+   * of that displaced value. Otherwise it will be a nullptr.
+   *
+   * /param key the key associated with the value to be stored
+   * /param value the value to store, and optionally the value displaced
+   */
+  bool
+  push_and_replace(uint64_t key, std::unique_ptr<T> & value)
```

Comment:
In terms of thread safety, I can see a situation where this function will return inconsistent state: if another thread pops the value that's being replaced after `did_replace` gets assigned but before the entry gets reassigned. Then the function would return `true` but value will be `nullptr`, which is the return value for when a value was not displaced.

However, a worse situation is if while this function is executing in Thread A, the context switches to Thread B and enters and finishes `push_and_replace`, which means that `head_` will get incremented. So if Thread A blocked after setting the new key but before setting the value or `in_use` bool, the mapping will be wrong.